### ISSUE TITLE
The .git directory is missing

### ISSUE THREAD

**Issue by reedy** [composer/composer#4717]
Given a clone of https://github.com/wikimedia/mediawiki-extensions-DonationInterface and also https://github.com/wikimedia/mediawiki-extensions-DonationInterface-vendor as the vendor repo of the parent repo...

If we just run composer update

```
reedy@ubuntu64-web-esxi:/var/www/wiki/mediawiki/extensions/DonationInterface$ php /var/www/wiki/mediawiki/composer.phar update
Loading composer repositories with package information
Updating dependencies (including require-dev)                                 
  - Removing predis/predis (v1.0.1)
  - Installing predis/predis (v1.0.3)
    Loading from cache

  - Updating coderkungfu/php-queue dev-master (4e3e6eb => c5a5200)


                                                                                                                                                                                    
  [RuntimeException]                                                                                                                                                                
  The .git directory is missing from /var/www/wiki/mediawiki/extensions/DonationInterface/vendor/coderkungfu/php-queue, see https://getcomposer.org/commit-deps for more informati  
  on                                                                                                                                                                                
                                                                                                                                                                                    


update [--prefer-source] [--prefer-dist] [--dry-run] [--dev] [--no-dev] [--lock] [--no-plugins] [--no-custom-installers] [--no-autoloader] [--no-scripts] [--no-progress] [--with-dependencies] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--ignore-platform-reqs] [--prefer-stable] [--prefer-lowest] [--] [<packages>]...


reedy@ubuntu64-web-esxi:/var/www/wiki/mediawiki/extensions/DonationInterface$ 
```

Ie we look at https://github.com/wikimedia/mediawiki-extensions-DonationInterface/blob/master/composer.json

And more specifically https://github.com/wikimedia/mediawiki-extensions-DonationInterface/blob/master/composer.json#L22 and https://github.com/wikimedia/mediawiki-extensions-DonationInterface/blob/master/composer.json#L31-L34

coderkungfu/php-queue being used, is from a fork git repo (the original is on packagist too). If someone makes the whole vendor repo, great, they have the .git folder. If they then commit it, and others try to update it, we get the error

If we blow away the whole repo, and then run composer update to rebuild it, it works fine

There should be a way to composer to recover from the situation, and re-clone the git repo.

I know committing the whole composer repo isn't usual practice, but it seems fairly widely done... And this issue is fairly well dcoumented all over the place https://www.google.co.uk/search?q=composer+The+.git+directory+is+missing&oq=composer+The+.git+directory+is+missing&aqs=chrome..69i57j0j69i60.2476j0j7&sourceid=chrome&es_sm=91&ie=UTF-8


Is there a way we can get composer to handle this situation better?

**Comment by olvlvl** [comment]
If you require the `.git` directory you should install/update with the `--prefer-source`, otherwise composer install packages from distribution file, it doesn't clone the repositories.

**Comment by bozerkins** [comment]
There, actually, is a PR, which would provide better handling of such situations.

https://github.com/composer/composer/pull/4878

**Comment by Seldaek** [comment]
Merged #4878 which indeed fixes this IMO.
